share/getters/shrex.go

```diff
@@ -97,33 +113,46 @@ func (sg *ShrexGetter) GetSharesByNamespace(
 	root *share.Root,
 	id namespace.ID,
 ) (share.NamespacedShares, error) {
-	var attempt int
+	var (
+		attempt int
+		err     error
+	)
 	for {
 		attempt++
-		select {
-		case <-ctx.Done():
-			return nil, fmt.Errorf("getter/shrex: %w", ctx.Err())
-		default:
-		}
-		peer, setStatus, err := sg.peerManager.Peer(ctx, root.Hash())
-		if err != nil {
-			log.Debugw("couldn't find peer", "datahash", root.String(), "err", err)
+		start := time.Now()
+		peer, setStatus, getErr := sg.peerManager.Peer(ctx, root.Hash())
+		if getErr != nil {
+			err = multierr.Append(err, getErr)
+			log.Debugw("couldn't find peer",
+				"datahash", root.String(),
+				"err", getErr,
+				"finished (s)", time.Since(start))
 			return nil, fmt.Errorf("getter/shrex: %w", err)
 		}
 
+		reqStart := time.Now()
 		reqCtx, cancel := ctxWithSplitTimeout(ctx, sg.minAttemptsCount-attempt+1, sg.minRequestTimeout)
-		nd, err := sg.ndClient.RequestND(reqCtx, root, id, peer)
+		nd, getErr := sg.ndClient.RequestND(reqCtx, root, id, peer)
 		cancel()
-		switch err {
+		switch getErr {
 		case nil:
 			setStatus(peers.ResultSuccess)
 			return nd, nil
 		case context.DeadlineExceeded:
-			log.Debugw("request exceeded deadline, trying with new peer", "datahash", root.String())
 		case p2p.ErrInvalidResponse:
 			setStatus(peers.ResultBlacklistPeer)
 		default:
 			setStatus(peers.ResultCooldownPeer)
 		}
+
+		if !ErrorContains(err, getErr) {
+			err = multierr.Append(err, getErr)
+		}
```

Comment:
I'm confused -- so if we get two peers who return to us an `ErrInvalidResponse` for example, we would only collect that err once?